arch/risc-v/src/common/crt0.c

```diff
@@ -88,6 +88,71 @@ static void sig_trampoline(void)
   );
 }
 
+/****************************************************************************
+ * Public Data
+ ****************************************************************************/
+
+#if defined(__cplusplus)
+extern "C"
+{
+#endif
+
+/* Linker defined symbols to .ctors and .dtors */
+
+  extern void (*_sctors)(void);
+  extern void (*_ectors)(void);
+  extern void (*_sdtors)(void);
+  extern void (*_edtors)(void);
+
+#if defined(__cplusplus)
+}
+#endif
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+#if defined(__cplusplus)
+extern "C"
+{
+#endif
+
+/****************************************************************************
+ * Name: exec_ctors
+ *
+ * Description:
+ *   Call static constructors
+ *
+ ****************************************************************************/
+
+  static void exec_ctors(void)
+  {
+    for (void (**ctor)(void) = &_sctors; ctor != &_ectors; ctor++)
+      {
+        (*ctor)();
+      }
+  }
+
+/****************************************************************************
+ * Name: exec_dtors
+ *
+ * Description:
+ *   Call static destructors
+ *
+ ****************************************************************************/
+
+  static void exec_dtors(void)
```

Comment:
it is already done by binfmt, please search CONFIG_BINFMT_CONSTRUCTORS.

Reply:
This way is also fine, and a quite widely used solution for static C++ ctors/dtors is to just put them into the crt files. There is no defined / required way to do this, so this way should be enough. For the kernel build using the linker defined symbols is the simplest way to do this.

Binfmt manually relocates ctors/dtors from the elf file to the binfmt structure. However, there is absolutely no reason to relocate ctors and dtors separately like binfmt does now. They are just initialized data, and data relocation does the work for us... I don't know why ctor/dtor handling was implemented in binfmt, as IMO it does not belong there. Maybe it is something needed by flat/protected ?

TLS ctors/dtors is another approach, but the kernel mode does not need this. I agree it would unify the implementations for FLAT/PROTECTED, but the scope of this PR is the kernel build.

Reply:
You need keep consider the flat and protected build when you develop kernel build feature.

Reply:
We can always do the big unification later. This does patch not break anything, and adds only a few lines of code that makes C++ work for build kernel.

I'm already doing one massive unification task, I'd prefer not to take another.

Reply:
Ok, I will do it in this week.